NativeOp.cpp

```diff
@@ -190,6 +190,78 @@ static void tf_cuda_sgemm(
 }
 
+template<typename T>
+static void tf_cuda_sgemm_batched(
+        OpKernelContext* context,
+        char transa, char transb,
+        int m, int n, int k,
+        const T* alpha_, const T** in_x, int lda,
+        const T** in_y, int ldb, const T* beta_,
+        const T** out,
+        int ldc,
+        int batchCount,
+        bool finalize_stream = 0) {
+    T alpha = *alpha_;
+    T beta = *beta_;
+// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/rnn/kernels/blas_gemm.cc
+#if GOOGLE_CUDA
+    int batch_size = batchCount;
+    typedef perftools::gputools::DeviceMemory<float> DeviceMemoryType;
+    std::vector<DeviceMemoryType> a_device_memory;
+    std::vector<DeviceMemoryType> b_device_memory;
+    std::vector<DeviceMemoryType> c_device_memory;
+    std::vector<DeviceMemoryType*> a_ptrs;
+    std::vector<DeviceMemoryType*> b_ptrs;
+    std::vector<DeviceMemoryType*> c_ptrs;
+    a_device_memory.reserve(batch_size);
+    b_device_memory.reserve(batch_size);
+    c_device_memory.reserve(batch_size);
+    a_ptrs.reserve(batch_size);
+    b_ptrs.reserve(batch_size);
+    c_ptrs.reserve(batch_size);
+
+    for (int64 i = 0; i < batch_size; ++i) {
+      a_device_memory.push_back(AsDeviceMemory(in_x[i]));
+      b_device_memory.push_back(AsDeviceMemory(in_y[i]));
+      c_device_memory.push_back(AsDeviceMemory(out[i]));
+      a_ptrs.push_back(&a_device_memory.back());
+      b_ptrs.push_back(&b_device_memory.back());
+      c_ptrs.push_back(&c_device_memory.back());
+    }
+
+    cudaStream_t cuda_stream = context->eigen_gpu_device().stream();
+
+    // cublasCreate, http://docs.nvidia.com/cuda/cublas/#cublascreate
+
+    auto dev_ctx = context->op_device_context();
+    perftools::gputools::Stream* dev_stream = dev_ctx->stream();
+    OP_REQUIRES(context, dev_stream, errors::Internal("No GPU stream available."));
+
+    bool blas_launch_status =
+        dev_stream
+             ->ThenBlasGemmBatched(get_transpose(transa), get_transpose(transb),
+                            (uint64)m, (uint64)n, (uint64)k, alpha, a_ptrs,
+                            lda, b_ptrs, ldb, beta, c_ptrs, ldc, batchCount)
+             .ok();
+    OP_REQUIRES(context, blas_launch_status, errors::Aborted("CuBlasGemm failed!"));
+
+    // The above call to ThenBlasGemmBatched allocates temporary memory on the GPU
+    // This memory is only freed once the stream is "finalized", which might not happen automatically
+    // Therefore, if the flag is set, we force a synchronization. This is bad for the performance, so it should be
+    // done as seldom as possible
+    if(finalize_stream) {
+        blas_launch_status =
+            dev_stream
+                ->BlockHostUntilDone()
+                .ok();
+        OP_REQUIRES(context, blas_launch_status, errors::Aborted("BlockHostUntilDone failed!"));
+    }
+#else  // GOOGLE_CUDA
+    context->SetStatus(errors::InvalidArgument("CuBlasGemm needs CUDA."));
+#endif  // GOOGLE_CUDA
+}
```

Comment:
`sgemm_batched` is needed for performance reasons. Usually, only one matrix matrix multiplication is needed at any time. But for the 2D-LSTM, we can compute a complete diagonal at once, so multiple operations can be performed in parallel. 

For CPU support, this could be changed to a sequence of multiple `sgemm` calls.

Reply:
I don't exactly understand. What is the matrix in `a_ptrs`, `b_ptr`, `c_ptr`, or better, what are their shapes?

Reply:
`a_ptrs` (as well as `b_ptrs` and ` c_ptrs`) wraps the input `in_x` (`in_y`, `out`).
This input is a sequence of matrices (for 2D-LSTM, one matrix for each element of the current diagonal). The number of these matrices is given by `batch_size` (which is equal to `batchCount`, I'll make it consistent).
In the `for` loop, they are wrapped using `AsDeviceMemory()` and added to `a_ptrs`.
The shape of the matrices is given by `m`, `n` and `k`: `in_x` has shape `(m, k)`, `in_y` has shape `(k, n)` and `out` has shape `(m, n)`.

Reply:
So, (m,k), (k,n) are the input shapes, (m,n) is the output shape. This is a pretty standard matrix multiplication, which should be possible via normal `sgemm`. I don't see what `sgemm_batched` can do better/faster than `sgemm`.

Reply:
But `sgemm` only performs one such multiplication at a time, right? We have to perform *multiple, independent* multiplications in parallel.